File: feature_engineering.R
# 清空工作空间
rm(list=ls())
# 安装R包
install.packages("ggplot2")
install.packages("factoextra")
install.packages("seqinr")
install.packages("dplyr")

# 加载13维氨基酸理化指标
index <- read.table("physicochemical_properties.txt")
# 将行名全部修改成大写
row.names(index) <- toupper(row.names(index))

# 计算距离矩阵
dt <- dist(scale(index[,c(8,7,10,11)]))

#===================系统聚类===================
# 系统聚类
hc <- hclust(dt,method = "average")
# 展示聚类谱系图
library(ggplot2);library(factoextra)
graph0 <- fviz_dend(hc, 
            k=6,                 ## 分成6类
            cex = 1,              ## 设置数据标签的字体大小
            horiz = T,                ## 水平摆放图形
            k_colors = 2:8,
            labels_track_height=0,
            color_labels_by_k=T,    ## 自动设置数据标签颜色
            lwd = 1               ## 设置分支的线宽
            ) +      ## 绘制矩形聚类图
  theme(axis.title.x = element_text(size = rel(1.5)),
        axis.text.x = element_text(size = rel(1.8)))  ## 修改坐标字体大小
graph0
# 加载序列数据
S <- seqinr::read.fasta(file.choose())
# 提取特征向量
get_tz <- function(m){
  lst <- list()
  for (i in 1:m) {
    n <- numeric(6)
    weizhi <- numeric(6)
    for (j in 1:length(S[[i]])) {
      s <- S[[i]][j]
      if (s %in% c("a","i","l","m","v","f")){
        n[1]=n[1]+1
        weizhi[1]=weizhi[1]+j
      }
      if (s %in% c("y")){
        n[2]=n[2]+1
        weizhi[2]=weizhi[2]+j
      }
      if (s %in% c("q","n","s","g","t","p","w")){
        n[3]=n[3]+1
        weizhi[3]=weizhi[3]+j
      }
      if (s %in% c("k","r","h")){
        n[4]=n[4]+1
        weizhi[4]=weizhi[4]+j
      }
      if (s %in% c("e","d")){
        n[5]=n[5]+1
        weizhi[5]=weizhi[5]+j
      }
      if (s %in% c("c")){
        n[6]=n[6]+1
        weizhi[6]=weizhi[6]+j
      }
    }
    
    lst[[i]] <- n/length(S[[i]])
    lst[[i]][7:12] <- weizhi/n
  }
  return(lst)
}

# 运行函数所需时间
system.time(tz <- get_tz(length(S)))
tz <- get_tz(length(S)) |> data.frame() |> t()

library(dplyr)
description <- seqinr::getAnnot(S) %>%  
  gsub(">","",.) # 将>批量转换为空格

row.names(tz) <- description
#===================相似性距离矩阵===================
# 计算距离矩阵
dt <- dist(scale(tz))
dt_show <- as.matrix(round(dt,3))
#===================系统聚类===================
# 系统聚类
hc <- hclust(dt,method = "average")

# 展示聚类谱系图
graph <- fviz_dend(hc, 
                   k=5,                 ## 分成5类
                   cex = 0.7,              ## 设置数据标签的字体大小
                   horiz=T,                ## 水平摆放图形
                   k_colors = c(2:4,6,7),
                   labels_track_height=5,
                   ylab = "Linkage distance",
                   color_labels_by_k=T,    ## 自动设置数据标签颜色
                   lwd = 1,               ## 设置分支的线宽
                   type = "rectangle") +      ## 绘制矩形聚类图
  theme(axis.title.y = element_text(size = rel(1)),
        axis.text.y = element_text(size = rel(1))) + ## 修改坐标字体大小
  labs(y="") ## 环形进化树不用显示y

graph

File: feature_engineering.py
# 清空工作空间
import gc
gc.collect()

# 安装需要的Python库
# pip install pandas numpy scipy matplotlib seaborn biopython

import pandas as pd
import numpy as np
from scipy.spatial.distance import pdist, squareform
from scipy.cluster.hierarchy import dendrogram, linkage
from scipy.stats import zscore
import matplotlib.pyplot as plt
from Bio import SeqIO
import os

os.chdir("D:\hierarchical-clustering")
# 加载13维氨基酸理化指标
index = pd.read_csv("physicochemical_properties.txt", sep="\t", index_col=0)
# 将行名全部修改成大写
index.index = index.index.str.upper()
# 计算距离矩阵之前进行标准化
scaled_index = index.iloc[:, [7, 6, 9, 10]].apply(zscore)
# 计算距离矩阵
dt = pdist(scaled_index, metric='euclidean')

#===================系统聚类===================
# 系统聚类
hc = linkage(dt, method='average')

# 展示聚类谱系图
plt.figure(figsize=(19, 11))
dendrogram(hc, labels=index.index, orientation='left', color_threshold=0)
plt.title('Hierarchical Clustering Dendrogram')
plt.xlabel('Distance')
plt.ylabel('Amino Acids')
plt.show()

# 加载序列数据
S = list(SeqIO.parse("influenza_A_virus.fasta", "fasta"))

# 提取特征向量
def get_tz(sequences):
    lst = []
    for seq in sequences:
        n = np.zeros(6)
        weizhi = np.zeros(6)
        for i, s in enumerate(seq.seq):
            if s in "AILMVF":
                n[0] += 1
                weizhi[0] += i + 1
            if s == "Y":
                n[1] += 1
                weizhi[1] += i + 1
            if s in "QNSGTPW":
                n[2] += 1
                weizhi[2] += i + 1
            if s in "KRH":
                n[3] += 1
                weizhi[3] += i + 1
            if s in "ED":
                n[4] += 1
                weizhi[4] += i + 1
            if s == "C":
                n[5] += 1
                weizhi[5] += i + 1
        feature_vector = np.concatenate([n / len(seq), weizhi / n])
        lst.append(feature_vector)
    return np.array(lst)

# 运行函数所需时间
import time
start_time = time.time()
tz = get_tz(S)
end_time = time.time()
print(f"运行时间: {round(end_time - start_time,4)} 秒")

# 转换为数据框
tz_df = pd.DataFrame(tz, index=[rec.description for rec in S])

#===================相似性距离矩阵===================
# 计算距离矩阵之前进行标准化
scaled_tz = tz_df.apply(zscore)
# 计算距离矩阵
dt = pdist(scaled_tz, metric='euclidean')
dt_show = squareform(dt)

#===================系统聚类===================
# 系统聚类
hc = linkage(dt, method='average')

# 展示聚类谱系图
plt.figure(figsize=(19, 11))
dendrogram(
    hc,
    labels=tz_df.index,
    orientation='left',
    color_threshold=0,
    leaf_font_size=8  # 调整标签字体大小
)
# 设置边距
plt.subplots_adjust(left=0.1, right=0.2, top=0.2, bottom=0.1)  # 调整边距
plt.title('Hierarchical Clustering Dendrogram')
plt.xlabel('Linkage Distance')
plt.ylabel('Sequences')
plt.tight_layout()
plt.show()
